airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
+        }
+
+        LOG.info("Database connection closed. Current version is: {}.", getHsqldbDatabaseVersion());
+    }
+
+    /**
+     * Checks if the database is correctly shut down
+     */
+    private boolean isHsqldbDatabaseCorrectlyShutdown() {
+
+        // The db/airsonic.data file may or may not exist (contains CACHED tables).
+        // The db/airsonic.backup file may or may not exist (not used during normal operation).
+
+        // The db/airsonic.properties file must exist (contains db properties).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".properties").exists()) {
+            LOG.warn("The db/airsonic.properties file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.script file must exist (contains db data from all tables except CACHED).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".script").exists()) {
+            LOG.warn("The db/airsonic.script file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.log file should not exist (contains db logs that are not supported in the upgrade).
+        if (new File(settingsService.getDefaultJDBCPath() + ".log").exists()) {
+            LOG.warn("The db/airsonic.log file exists. The database may not be shut down correctly.");
+            return false;
+        }
+        // The db/airsonic.lck file should not exist (lock file, indicates another running process or unclean shutdown).
+        if (new File(settingsService.getDefaultJDBCPath() + ".lck").exists()) {
+            LOG.warn("The db/airsonic.lck file exists. The database may not be shut down correctly.");
+            return false;
+        }
+
+        return true;
+    }
+
+    private void upgradeHsqldbDatabaseSafely() {
+        try {
+            // Check that it is safe to do the upgrade
+            if (!isHsqldbDatabaseCorrectlyShutdown()) {
+                LOG.warn(
+                    "Database has not been shut down properly. " +
+                    "Airsonic should be stopped properly before attempting to upgrade the database.");
+            }
```

Comment:
hmmm, i don't understand this logic.
- the bean's instantiated
- postconstruct checks if upgrade is needed, then calls this method
- why should the db have been shutdown at this point? didn't the application just start?
- wouldn't this generate a warn for everyone?

Reply:
If  check for unexpected startups other than embedded mode, not harmful logic.
It may be unnecessary if the case can be avoid by the framework locking mechanism.
The check logic that can be useful against the legacy framework implementation of hsql is implemented...?

Reply:
Hmmm, not sure I understand. What other startups other than embedded mode?

While it's not harmful logic, I agree, because it's just issuing warnings, it (a) slows down startup due to file checks, and (b) is an unnecessary check: this condition always evaluates to true. We don't even need to check.

It always is true because, unless there was manual intervention involved, the database isn't shutdown, there is no logic prior to this I see that shuts it down in the code, and there is no reasonable expectation for it to have been shut down automatically. In fact, since the application is starting, the db should be starting up and I'd be worried if it _was_ shutdown.

Reply:
The condition doesn't always evaluates to true -- in fact in 99% cases at this point in the startup sequence it should be false. This bean should start before any other database operation has been run, so at this point the db isn't started yet and should be in shut down state.

The scenario I want to check is that for some reason the previous Airsonic version (before the upgrade) is not shut down properly (either power loss, timeout that kills the database, etc.). This leaves a log file behind, which is not supported at all by the upgrade process (this means that such users _will_ lose data).

Reply:
I don't think that's true at all. There are three issues here:

- > at this point the db isn't started yet and should be in shut down state.

You're thinking that the db is already shutdown when you start an application. While that may be true of recent versions (which I suspect you're using for testing), _all_ versions of subsonic/libresonic/airsonic prior to recent times _didn't_ shutdown the DB at termination, to my recall (termination of threads to the DB doesn't shut it down unless you specify in the connection parameters). Every person jumping from prior to whenever the shutdown hook was introduced (which was just this year/recently?) will have this evaluation true. And so will everyone who migrates from subsonic to airsonic. And so will every person who just shuts down their JVM without a proper SIGTERM signal (which is most cloud deployments, anyone who issues a `docker kill` to stop their container etc). So, by definition, you're only counting an extremely narrow subset of people who shutdown their embedded hsql dbs properly last time (that is, only people who have the most latest version of airsonic, not using docker kills), and excluding most of them who have no reasonable expectation of a shutdown db. And even for that narrow segment:

- > This bean should start before any other database operation has been run 

You can't make that assertion. There are no guarantees when the connection pool instantiates a connection to the database. That's not under user control. While you may control user-defined DB accesses, connections pools might eagerly establish connections just to check if they're alive etc _prior_ to a user command execution. You're getting lucky that your connection pool isn't establishing a connection to the DBs prior to this check, and should not rely on that behavior.

Both the issues above highlight that minus a very specific subset of installs and ideal situations based on luck or undocumented pool behaviors, the DB won't be shutdown and shouldn't be expected to. And even then, it doesn't matter because:

- > The scenario I want to check is that for some reason the previous Airsonic version (before the upgrade) is not shut down properly (either power loss, timeout that kills the database, etc.). This leaves a log file behind, which is not supported at all by the upgrade process (this means that such users will lose data).

Your check here won't stop that upgrade, I'm not sure what you mean by "not supported". Arguably _nothing_ you do here will stop an upgrade (unless you throw exceptions preventing the application from starting up).

Let's say this condition evaluates to false, and we skip this conditional. Your DB is still getting upgraded at next SQL instruction.

We're just trying to make a backup. You're not actually doing the upgrade, that's done by the driver automatically, there's no safe or unsafe. That's happening no matter what.

Reply:
> There are no guarantees when the connection pool instantiates a connection to the database.

That's, I think, exactly my point when I said, earlier, "one thing I'm still uneasy about is that the controlled upgrade procedure requires that nothing opens a connection to the data source while we're doing it".

> So the only thing we can do is:
>    Make a backup prior to first connection (do checking of db and driver version match to see if its needed)
>    Connect (for upgrade)
>    Compact db

I'll update the PR and do that.

Reply:
Done.